Quantify wind speed shift

samuelwnaylor opened the pull request.

samuelwnaylor committed:
refactor: move ops curve shift calcs and reduce complexity of functions

- Moves Operational Curve shift calculations into their own module
- Refactor into separate curve specific functions
  (with underlying more generic private functions)

This change improves readability and ease of adding further curve shift
calculations.

samuelwnaylor committed:
feat: add function to quantify wind speed shift

In similar way that power curve, rpm and pitch shifts are already
calculated, this change adds wind speed shift to the ops curve shift
module

samuelwnaylor committed:
refactor: log the curve shift value sign in the warning message

Display the sign (+/-) of the curve shift value in the log warning
message for clarity of the actual shift rather than just displaying the
abs() value that is used when determining whether to raise a log warning

aclerc commented on wind_up/main_analysis.py:
```diff
@@ -365,78 +366,6 @@ def yaw_offset_results(
-            msg = f"check_for_ops_curve_shift {wtg_name} post_df missing required column {req_col}"
-            result_manager.warning(msg)
-            return results_dict
-    pre_dropna_df = pre_df.dropna(subset=[scada_ws_col, pw_col, pt_col, rpm_col]).copy()
-    post_dropna_df = post_df.dropna(subset=[scada_ws_col, pw_col, pt_col, rpm_col]).copy()
```
I am not completely sure but I think this dropna logic is missing in the new `ops_curve_shift.py`

samuelwnaylor replied:
> I am not completely sure but I think this dropna logic is missing in the new `ops_curve_shift.py`

This logic is here now https://github.com/resgroup/wind-up/blob/173f074d25a61c633fd41f58a711d85142f22383/wind_up/ops_curve_shift.py#L71

I've added NaN handling to the tests too...

[NaN included in df](https://github.com/resgroup/wind-up/blob/173f074d25a61c633fd41f58a711d85142f22383/tests/test_ops_curve_shift.py#L23) and the [assertion in a test](https://github.com/resgroup/wind-up/blob/173f074d25a61c633fd41f58a711d85142f22383/tests/test_ops_curve_shift.py#L112)

aclerc reviewed (changes requested):
I am seeing some differences in integration tests, will email you with further detail. I might have found the reasons (see comments) but not totally sure. Apologies there were not unit tests for the original `check_for_ops_curve_shift` function you are replacing

samuelwnaylor committed:
address pr comments

samuelwnaylor committed:
fix: set required columns to be all curve columns

Required columns are necessary for all curve plot types. The required
columns for any curve plot are wind_speed, power, pitch and rpm

samuelwnaylor committed:
run smarteole_example.ipynb

samuelwnaylor committed:
fix: correct the if in

aclerc commented on wind_up/ops_curve_shift.py:
```diff
@@ -0,0 +1,239 @@
+    post_df["expected_y"] = np.interp(post_df[conf.x_col], mean_curve["x_mean"], mean_curve["y_mean"])
+    mean_df = post_df.mean()
+
+    if conf.name in [CurveTypes.PITCH.value, CurveTypes.PITCH]:
```
It seems like it would be better if there were exactly 1 way to identify if the curve shift calculation is for pitch. What about give conf a curvetype attribute which is type CurveTypes so you would have `if conf.curve_type == CurveTypes.PITCH:`

samuelwnaylor replied:
Since `CurveTypes` inherits from `str` and `Enum`, then we could simply it to `if conf.name == CurveTypes.PITCH`.

aclerc reviewed (changes requested):
Maybe I am misunderstanding the commit history but it seems there are no test case changes associated with this fix https://github.com/resgroup/wind-up/pull/42/commits/6045c39a3cf5a2c0ae77f4ad5ab57a2546c40bbd. If that's the case then I think a test relevant to this logic needs adding, ie at least one test should fail before this commit and pass after it

samuelwnaylor commented:
> Maybe I am misunderstanding the commit history but it seems there are no test case changes associated with this fix [6045c39](https://github.com/resgroup/wind-up/commit/6045c39a3cf5a2c0ae77f4ad5ab57a2546c40bbd). If that's the case then I think a test relevant to this logic needs adding, ie at least one test should fail before this commit and pass after it

The tests were 

> Maybe I am misunderstanding the commit history but it seems there are no test case changes associated with this fix [6045c39](https://github.com/resgroup/wind-up/commit/6045c39a3cf5a2c0ae77f4ad5ab57a2546c40bbd). If that's the case then I think a test relevant to this logic needs adding, ie at least one test should fail before this commit and pass after it

The test case did not change because by chance the test case had `conf.name = "pitch"` and `conf.y_col = "pitch"`. It was only when running analysis for an actual wind farm that the results were different prior to the fix in commit https://github.com/resgroup/wind-up/commit/6045c39a3cf5a2c0ae77f4ad5ab57a2546c40bbd.

I'll add different strings for `conf.name` and `conf.y_col` in the test case

samuelwnaylor committed:
address pr comments
